src/parser/error.rs

```diff
@@ -19,4 +18,11 @@ pub enum ParseError {
     Lexical(LexicalError),
     /// Unbound type variable(s)
     UnboundTypeVariables(Vec<Ident>, RawSpan),
+    /// Illegal record literal in the uniterm syntax (see RFC002). In practice, this is a record
+    /// mixing a tail and non-type constructs.
+    InvalidUniRecord(
+        RawSpan, /* non-type construct position */
+        RawSpan, /* tail position */
+        RawSpan, /* whole record position */
+    ),
```

Comment:
I've seen this before…

Do we need the error to be defined both in `error` and `parser::error`?

Same remarks as previously on the docstring.

Reply:
`error::ParseError` is a superset of `parser::error::ParseError`. The latter are custom errors that we can raise from within LALRPOP: it includes lexing errors (because we use a custom lexer), and now the `InvalidUniRecord`. Then LALRPOP can raise those errors or generic ones (unexpected end of file, unexpected token, etc.). On level up, `error::ParseError` also includes additional parsing errors when deserializing e.g. json files, and also wrap LALRPOP generic and custom errors with additional information, such as the file identifier.

For modeling errors, I've often missed structural typing, something like OCaml polymorphic variants. Right now we duplicate quite a few constructors to model the inclusion of some errors in a larger error set but with additional information. One alternative would be to reuse the previous level when possible, adding data if needed, something like `LALRPOPError(lalrpop::SomeErrorType, FileId)`. One disadvantage is that in this precise case, the error is much less nice to handle because `lalrpop::SomeErrorType` has usually a different and LALRPOP-specific representation. In practice the only part consuming errors is the one turning them into a diagnostics, so that's maybe not a big deal.